Merge of opal's fork

elia opened the pull request:
> eregontp: @opalrb Nice! I would be interested to know how you guys use RubySpec :) Definitely open an issue if we can make it easier for you.

– https://twitter.com/eregontp/status/604710586300133376

I don't expect this PR to be merged right away, rather I'm opening it to check & fix all the little things that will make it ok for a merge.

cc @eregon @opal/core

eregon commented on core/basicobject/basicobject_spec.rb:
```diff
@@ -76,7 +76,7 @@ def test_method() :test end
     obj.instance_variable_get(:@test).should == :value
 
-    obj.send(:hash).should be_an_instance_of(Fixnum)
+    lambda { obj.send(:hash) }.should_not raise_error
```
Would it work for Opal if it was
`obj.send(:hash).should be_an_instance_of(Integer)` ?
That would be a more precise assertion.

elia replied:
No, because unfortunately we use object literals as out hash table implementation, thus our keys are strings.
The intent was to merely ensure the method is there and "works".

As the test name is "contains Kernel methods when including Kernel" at the time I thought it made sense, but can be easily moved to an "unsupported" block.

eregon replied:
You are right, the spec is just about the existence of `#hash`.
A `should respond_to(:hash)` would make a lot more sense then.

eregon commented on core/array/hash_spec.rb:
```diff
@@ -11,29 +11,33 @@
+      rec.hash.should == [{:x => rec}].hash
+      # Like above, this is because rec.eql?([{:x => rec}])
+    end
   end
```
I think the changes in this file are just the result of a bad merge. They just have additional `ruby_bug` but on unsupported versions so they were removed.

eregon commented on spec_helper.rb:
```diff
@@ -48,10 +48,7 @@
-CODE_LOADING_DIR = defined?(require_relative) ?
-                     File.realpath(dir, __FILE__) :
-                     File.expand_path(dir, __FILE__)
+CODE_LOADING_DIR = File.expand_path "../fixtures/code", __FILE__
```
Is the call to `defined?(require_relative)` problematic?

elia replied:
it's probably that we're missing `realpath`

eregon replied:
Also checking for `File.respond_to? :realpath` would be better indeed.

elia replied:
I think the check on require_relative comes from ruby 1.8 support and seems like the difference between `realpath` and `expand_path` is just that realpath ensures the file actually exists, but I might be wrong here…

vais replied:
@elia actually if I remember correctly the issue was not that we are missing `realpath` (it's implemented in nodejs), but it was returning a path that was missing one (the first) component. I tried to figure it out and gave up. I think it's something to do with how nodejs sees current working directory. I gave up mostly because I didn't see the point of that code in the first place. @eregon, pardon my ignorance, what's the rationale for branching on `defined?(require_relative)` here? I.e. what problem does that solve as opposed to simply using `File.expand_path "../fixtures/code", __FILE__` unconditionally?

eregon replied:
The difference is `realpath` resolves symlinks while `expand_path` won't. I am sure of the actual reason though. It comes from 6690a69a7a608f3b4e47fdef70d8776ad9a372c9.

eregon replied:
It seems to be that since we have `require_relative`, features in $LOADED_FEATURES are expanded with realpath (their symlinks are resolved) while they used to not be before `require_relative`.
This in turn probably causes spec failures if executing rubyspec and the current directory contains symlinks.
@nobu Could you confirm I understand your commit correctly?

vais replied:
@eregon ooh, I see now. @elia could use your help there :sweat_smile: Basically, as I mentioned above, realpath works, but returns a shorter path than expected, making the require fail.

eregon commented on language/regexp/character_classes_spec.rb:
```diff
@@ -96,9 +96,11 @@
+      ('/[[:ascii:]]/').match("\u{80}").should be_nil
+      ('/[[:ascii:]]/').match("\u{9898}").should be_nil
+    end
   end
```
Why not use a tag/exclude in this case since the whole example is not compliant?

elia replied:
Not sure about this but some stuff needs to be syntactically in an excluded block because [it's skipped at compile time](https://github.com/opal/opal/blob/master/lib/mspec/opal/special_calls.rb)

@vais will know better I think

vais replied:
@elia @eregon that's right, unfortunately in this case filtering it out happens too late in the process, the code simply fails to compile.

eregon replied:
I see, although in this particular case both arguments are String (which is a bug, fixed on master) so I would not expect that to be problematic at compile time.

vais replied:
@eregon hehe, it's javascript, *expect the unexpected*(tm) :stuck_out_tongue_winking_eye: Pasting `s = "\u{80}";` into Chrome console:
```
s = "\u{80}";
VM106:2 Uncaught SyntaxError: Unexpected token ILLEGAL
    at Object.InjectedScript._evaluateOn (<anonymous>:895:140)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:828:34)
    at Object.InjectedScript.evaluate (<anonymous>:694:21)
```

eregon replied:
:smile:
Would translating that to "\u0080" make sense?

eregon commented on core/string/shared/succ.rb:
```diff
@@ -50,6 +50,7 @@
     "9Z99z99Z".send(@method).should == "10A00a00A"
 
     "ZZZ9999".send(@method).should == "AAAA0000"
+    "/[]9999".send(@method).should == "/[]10000"
```
I do like new specs!
I see you even have https://github.com/opal/opal/blob/master/spec/opal/core/date_spec.rb for instance, which would be very interesting to merge if it is compatible with MRI.

eregon commented:
Hello!

I see you use a [custom file](https://github.com/opal/opal/blob/master/spec/rubyspecs) instead of a MSpec config to list specs to run.
And a [custom set of excludes/filters](https://github.com/opal/opal/tree/master/spec/filters) instead of MSpec tags.
Is there any particular reason? Would it be because filesystem interactions are difficult or executing some parts of MSpec would require additional efforts?

I made a few comments on the diff, but for the rest I see it is mainly `not_supported_on` guards.
In general we try to have less guards like these as it impacts readability, makes it harder to maintain and in the end makes some parts of the spec just ignored under some implementations, which is not the goal of RubySpec.
Ideally, tags would be enough for excluding specs which are not supported, but of course that excludes a whole example (`it`). One solution would be to break the assertions so there is only a few `should` per examples. This might be good for some cases (like for `equal?` with strings) but might also be very cumbersome for some other cases (like Regexp options).

eregon commented:
What features are unsupported by Opal?
I see http://opalrb.org/docs/unsupported_features/ and from the specs here also some Regexp options and Bignum?

elia commented:
@eregon yes 😓 docs need some love

The real list is made of the unsupported filters
https://github.com/opal/opal/tree/master/spec/filters/unsupported

plus the stuff commented in the specs list
https://github.com/opal/opal/blob/master/spec/rubyspecs

that is not part of the bug list
https://github.com/opal/opal/tree/master/spec/filters/bugs

hope it makes sense…

eregon commented:
`bugs` and `unsupported` are combined together as an example filter and the specs list filters files, right?

elia commented:
Yes the difference is just to help us track what we intend to fix and what we can't do anything about (e.g. threads/encoding/tainting)

elia commented:
I think I addressed all comments, removing the WIP mark, let me know if there's anything else that needs to be done for the PR to be merged.

I left the date_spec update out on purpose as I'd like to prepare a proper PR for it keeping @adambeynon as the spec author.

elia commented:
One last note, 

> You are right, the spec is just about the existence of #hash.
> A should respond_to(:hash) would make a lot more sense then.

I've updated the spec to use respond_to but I noticed that this way it's not calling the method anymore. Not necessarily a problem, but still worth being aware of the change.

eregon commented:
The last commit causes failures, so it's probably better to leave it as:
```
lambda { obj.hash }.should_not raise_error
```
(The send should not be needed, right?)

elia commented:
ha :) just force pushed with

```rb
    obj.respond_to?(:hash).should == true
```

elia commented:
@eregon I can revert it if you want

eregon commented:
@elia That's good too, thank you for your work!

I am still looking for a better solution than all these guards.
I think some examples can be split easily and the other are mostly compile-time issues, right?

For the compile-time problems, would it be possible to accept the syntax and compile unsupported things to something that would just raise a "NotImplementedError" or "NotSupportedError"? There is something similar in JRuby+Truffle for instance.
If you want to keep a hard error at compile-time for normal usage, maybe it would make sense to have an extra flag when running RubySpec meaning "do not fail hard with unsupported syntax and compiler stuff".
But this delayed error might actually prove useful to load third-party code and only error out on usage of unsupported stuff.

eregon commented on core/integer/even_spec.rb:
```diff
@@ -9,10 +9,12 @@
     1.even?.should be_false
     2.even?.should be_true
 
```
This case for instance, would nicely be split in
`Integer#even? for Fixnum returns ...` and
`Integer#even? for Bignum returns ...`.

elia commented:
@eregon yes for some stuff we have that and I think it's the way to go long term, that being said I don't think I'll be able to tackle this change any time soon.

cc @vais

eregon commented:
@elia Right, fair enough. Given the number of exclusions is not too big I think this is OK to merge.
Do you mind if I rebase the commits on top of master to avoid multiple merges?

eregon commented on core/regexp/source_spec.rb:
```diff
@@ -7,13 +7,17 @@
-  it "has US-ASCII encoding when created from an ASCII-only \\u{} literal" do
-    re = /[\u{20}-\u{7E}]/
-    re.source.encoding.should equal(Encoding::US_ASCII)
+  not_compliant_on :opal do
```
I am trying to remove `non_compliant_on` guards in general, so I'll switch those to `not_supported_on` (which is the intended meaning here I believe).

eregon commented:
All changes are merged as of 388a4adbe1b05b283b85dea048f8319ccc2b0b54.
I cherry-picked the commits on top of master, removed the ones that reverted each other and adapted a few commit messages.
Thank you for your fixes and let's use ruby/rubyspec for Opal!

hsbt commented:
@eregon @elia :+1: Good work

elia commented:
Great! Thanks @eregon!
